contracts/marketplace/src/lib.rs

```diff
@@ -0,0 +1,274 @@
+#![no_std]
+use soroban_sdk::{
+    contract, contracterror, contractimpl, contracttype, panic_with_error,
+    token::{self},
+    Address, Env, Map, Symbol,
+};
+
+#[derive(Clone, Copy)]
+#[contracttype]
+enum DataKey {
+    Admin = 1,
+    Percentage = 2,
+    AlreadyInitialized = 3,
+    Assets = 4,
+    Token = 5,
+}
+
+#[contracterror]
+#[derive(Clone, Debug, Copy, Eq, PartialEq, PartialOrd, Ord)]
+#[repr(u32)]
+pub enum Error {
+    AlreadyInitialized = 1,
+    InvalidPercentage = 2,
+    InvalidAssetPrice = 3,
+    BalanceTooLow = 4,
+    AssetNotListed = 5,
+    InvalidAuth = 6,
+}
+
+#[contracttype]
+#[derive(Clone)]
+pub struct Asset {
+    owner: Address,
+    price: i128,
+    listed: bool,
+}
+
+#[contract]
+pub struct MarketplaceContract;
+
+#[contractimpl]
+impl MarketplaceContract {
+    /// Initialize the contract with the admin's address and a percentage for the admin fee.
+    pub fn init(env: Env, token: Address, admin: Address, percentage: i128) {
```

Comment:
Should we check in all other contract functions that this method was invoked (initilized) ?

Reply:
But it would just panic right?